lerobot/common/policies/rollout_wrapper.py

```diff
@@ -0,0 +1,294 @@
+                if k.startswith("observation")
+            }
+
+            # Forget any observations we won't be needing any more.
+            self._invalidate_obsolete_observations()  # TODO(now)
+
+        # Run inference.
+        device = get_device_from_parameters(self.policy)
+        observation_sequence_batch = {
+            key: observation_sequence_batch[key].to(device, non_blocking=True)
+            for key in observation_sequence_batch
+        }
+        actions = self.policy.run_inference(observation_sequence_batch).cpu()  # (batch, seq, action_dim)
+
+        # Update action cache.
+        with self._thread_lock:
+            self._action_cache.update(
+                {
+                    observation_timestamp_ms + i * self.period_ms: action
+                    for i, action in enumerate(actions.transpose(1, 0))
+                }
+            )
+
+    def _get_contiguous_action_sequence_from_cache(self, first_action_timestamp_ms: float) -> Tensor | None:
+        with self._thread_lock:
+            action_cache = deepcopy(self._action_cache)
+        if len(action_cache) == 0:
+            return None
+        action_cache_timestamps_ms = torch.tensor(sorted(action_cache))
+        action_timestamps_ms = (
+            torch.arange(0, action_cache_timestamps_ms.max() + self.period_ms, self.period_ms)
+            + first_action_timestamp_ms
+        )
+        dist = torch.cdist(
+            action_timestamps_ms.unsqueeze(-1).float(),
+            action_cache_timestamps_ms.unsqueeze(-1).float(),
+            p=1,
+        ).int()
+        min_, argmin_ = dist.min(dim=1)
+        if min_[0] > self.timestamp_tolerance_ms:
+            return None
+        # Get contiguous sequence of argmins_ starting from 0.
+        where_jump = torch.where(argmin_.diff() != 1)[0]
+        if len(where_jump) > 0:
+            argmin_ = argmin_[: where_jump[0] + 1]
+        # Retrieve and stack the actions.
+        action_sequence = torch.stack(
+            [action_cache[ts.item()] for ts in action_cache_timestamps_ms[argmin_]],
+            dim=0,
+        )
+        return action_sequence
+
+    def provide_observation_get_actions(
+        self,
+        observation_batch: dict[str, Tensor],
+        observation_timestamp: float,
+        first_action_timestamp: float,
+        strict_observation_timestamps: bool = False,
+        timeout: float | None = None,
+    ) -> Tensor | None:
+        """Provide an observation and get an action sequence back.
+
+        This method does several things:
+        1. Accepts an observation with a timestamp. This is added to the observation cache.
+        2. Runs inference either synchronously or asynchronously.
+        3. Returns a sequence of actions starting from the requested timestamp (from a cache which is
+           populated by inference outputs).
+
+        If `timeout` is not provided, inference is run synchronously. If `timeout` is provided, inference runs
```

Comment:
I think this timeout param is confusing. If I understand it correctly, the whole point of the PolicyRollout is because inference is likely to take longer than 1/fps (e.g. 0.1s). As such, when this method decides to run inference, there should be no expectation that inference completes within 0.1s and returns an action sequence. 

I would therefore propose that when inference is called, there should be no "attempt to wait for inference to complete, within the bounds of the `timeout` parameter." It should just run inference asynchronously, and continue to return the existing action_sequence. However, the timeout parameter should only check that the asynchronous inference process runs within the specified timeout period (e.g. 0.5s), and throws an error / warning if inference duration exceeds that timeout value. This check is merely to ensure that the action buffer doesn't get drawn down to 0 before a new set of actions are added to the buffer.